engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java

```diff
@@ -843,72 +846,103 @@ protected boolean handleDisconnectWithInvestigation(final AgentAttache attache,
+                    }
+                    else {
+                        _alertMgr.sendAlert(AlertManager.AlertType.ALERT_TYPE_HOST, host.getDataCenterId(), host.getPodId(), hostShortDesc + " is down", "Host " + hostDesc + " is down.");
+                    }
+                    event = Status.Event.HostDown;
+                    removeAgent = true;
+                } else if (determinedState == Status.Alert){
+                    // This is likely a Console Proxy or Secondary Storage VM
+                    s_logger.error("Investigation found " + hostShortDesc + " in state: " + determinedState);
+                    _alertMgr.sendAlert(AlertManager.AlertType.ALERT_TYPE_HOST, host.getDataCenterId(), host.getPodId(), "Investigation found " + hostShortDesc + " in " + determinedState, "Investigation found " + hostShortDesc + " in " + determinedState + " state.  Event is " + event);
                 } else {
-                    // if we end up here we are in alert state, send an alert
-                    final DataCenterVO dcVO = _dcDao.findById(host.getDataCenterId());
-                    final HostPodVO podVO = _podDao.findById(host.getPodId());
-                    final String podName = podVO != null ? podVO.getName() : "NO POD";
-                    final String hostDesc = "name: " + host.getName() + " (id:" + host.getId() + "), availability zone: " + dcVO.getName() + ", pod: " + podName;
-                    _alertMgr.sendAlert(AlertManager.AlertType.ALERT_TYPE_HOST, host.getDataCenterId(), host.getPodId(), "Host in ALERT state, " + hostDesc,
-                                    "In availability zone " + host.getDataCenterId() + ", host is in alert state: " + host.getId() + "-" + host.getName());
+                    // Determined state was not Up, Disconnected, Down, or Alert.  To catch anything else create alert
+                    s_logger.error("Investigation found " + hostShortDesc + " in unhandled state: " + determinedState);
+                    _alertMgr.sendAlert(AlertManager.AlertType.ALERT_TYPE_HOST, host.getDataCenterId(), host.getPodId(), "Investigation found " + hostShortDesc + " in " + determinedState, "Investigation found " + hostShortDesc + " in unhandled state: " + determinedState);
                 }
             } else {
-                s_logger.debug("The next status of agent " + host.getId() + " is not Alert, no need to investigate what happened");
+                s_logger.info("The next status of host " + hostShortDesc + " is " + nextStatus + ", no need to investigate what happened");
             }
         }
-        handleDisconnectWithoutInvestigation(attache, event, true, true);
-        host = _hostDao.findById(hostId); // Maybe the host magically reappeared?
+
+        handleDisconnectWithoutInvestigation(attache, event, true, removeAgent);
+        host = _hostDao.findById(hostId); // We may have transitioned the status - refresh
```

Comment:
protected boolean handleDisconnectWithInvestigation(final AgentAttache attache, Status.Event event) just grew from 85 to 118 lines. Can you please make it a bit more modular?

Reply:
Yea, that sounds good